iroha/src/asset.rs

```diff
@@ -21,68 +24,218 @@ impl Asset {
+            }
+            Component::BigQuantity(component) => {
+                if let Component::BigQuantity(other_component) = other {
+                    *component += other_component
+                }
+            }
+            Component::KeyValueComponent(component) => {
+                if let Component::KeyValueComponent(other_component) = other {
+                    *component += other_component
+                }
+            }
+        }
+    }
+}
 
-    /// Feels the `Asset` with the quantity.
-    pub fn with_quantity(mut self, quantity: u128) -> Self {
-        self.quantity = quantity;
-        self
+impl SubAssign for Component {
+    fn sub_assign(&mut self, other: Self) {
+        match self {
+            Component::Quantity(component) => {
+                if let Component::Quantity(other_component) = other {
+                    *component -= other_component
+                }
+            }
+            Component::BigQuantity(component) => {
+                if let Component::BigQuantity(other_component) = other {
+                    *component -= other_component
+                }
+            }
+            Component::KeyValueComponent(component) => {
+                if let Component::KeyValueComponent(other_component) = other {
+                    *component -= other_component
+                }
+            }
+        }
     }
+}
 
-    /// Constructor of the `Mint<Asset, u128>` Iroha Special Instruction.
-    pub fn mint(&self, object: u128) -> Mint<Asset, u128> {
+impl QuantityComponent {
+    /// Constructor of the `Mint<Component, u32>` Iroha Special Instruction.
+    pub fn mint(&self, object: u32) -> Mint<QuantityComponent, u32> {
         Mint {
             object,
             destination_id: self.id.clone(),
         }
     }
 }
 
-/// Identification of an Asset. Consists of Asset's name, Domain's name and Account's name.
+impl From<QuantityComponent> for Component {
+    fn from(component: QuantityComponent) -> Component {
+        Component::Quantity(component)
+    }
+}
+
+impl AddAssign for QuantityComponent {
+    fn add_assign(&mut self, other: Self) {
+        self.quantity += other.quantity;
+    }
+}
+
+impl AddAssign for BigQuantityComponent {
+    fn add_assign(&mut self, other: Self) {
+        self.quantity += other.quantity;
+    }
+}
+
+impl AddAssign for KeyValueComponent {
+    #[allow(clippy::redundant_clone)]
+    fn add_assign(&mut self, other: Self) {
+        self.store.append(&mut other.store.clone());
+    }
+}
+
+impl SubAssign for QuantityComponent {
+    fn sub_assign(&mut self, other: Self) {
+        self.quantity -= other.quantity;
+    }
+}
+
+impl SubAssign for BigQuantityComponent {
+    fn sub_assign(&mut self, other: Self) {
+        self.quantity -= other.quantity;
+    }
+}
+
+impl SubAssign for KeyValueComponent {
+    fn sub_assign(&mut self, other: Self) {
+        for key in other.store.keys() {
+            self.store.remove(key);
+        }
+    }
+}
```

Comment:
I am not sure if we should implement `SubAssign` for `KeyValueComponent`. As the operation itself seems a bit misleading here. From the users stand point it is not obvious whether it will match full pairs or just keys.

Reply:
True and use cases for quantities and KVs are very different, so there is not a need to structure this in a semi-generic complex way. The same asset can also have a qty and a KV store for data.

Reply:
ok